File: backend/api/schemas/query_runs.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from persistence.models import QueryRun


class QueryRunCreate(BaseModel):
    datasource_id: UUID
    question: str = Field(min_length=3, max_length=1_000)


class QueryRunError(BaseModel):
    code: str
    message: str


QueryRunStatus = Literal[
    "received",
    "retrieve_context",
    "generate_query",
    "validate_query",
    "execute_query",
    "repair_query",
    "verify_result",
    "select_visualization",
    "completed",
    "clarification_required",
    "out_of_scope",
    "blocked",
    "failed",
]


class QueryRunSummary(BaseModel):
    run_id: UUID
    datasource_id: UUID
    datasource_name: str
    question: str
    status: QueryRunStatus
    row_count: int | None
    duration_ms: int | None
    repair_count: int
    provider_call_count: int
    visualization_type: str | None
    error_code: str | None
    created_at: datetime


class QueryRunPage(BaseModel):
    items: list[QueryRunSummary]
    limit: int
    total: int
    next_cursor: str | None
    has_more: bool


class QueryRunResponse(BaseModel):
    run_id: UUID
    datasource_id: UUID
    question: str
    status: str
    query_type: str
    generated_query: dict[str, Any] | None
    validation: dict[str, Any]
    result: dict[str, Any] | None
    repair_count: int
    provider_call_count: int
    visualization_type: str | None
    clarification_suggestions: list[str]
    warnings: list[str]
    error: QueryRunError | None
    created_at: datetime


class QueryTraceResponse(BaseModel):
    run_id: UUID
    status: str
    trace: list[dict[str, Any]]


def query_run_response(run: QueryRun) -> QueryRunResponse:
    raw_suggestions = run.validation_result.get("clarification_suggestions", [])
    suggestions = (
        [str(item) for item in raw_suggestions] if isinstance(raw_suggestions, list) else []
    )
    return QueryRunResponse(
        run_id=run.id,
        datasource_id=run.datasource_id,
        question=run.question,
        status=run.status,
        query_type=run.query_type,
        generated_query=run.generated_query or None,
        validation=run.validation_result,
        result=run.result_json,
        repair_count=run.repair_count,
        provider_call_count=run.provider_call_count,
        visualization_type=run.visualization_type,
        clarification_suggestions=suggestions,
        warnings=run.warnings,
        error=(
            QueryRunError(code=run.error_code, message=run.error_message or "Query run failed")
            if run.error_code is not None
            else None
        ),
        created_at=run.created_at,
    )

File: frontend/lib/query-runs.ts
import { apiRequest } from "@/lib/api-client";

export type QueryRunStatus =
  | "received"
  | "retrieve_context"
  | "generate_query"
  | "validate_query"
  | "execute_query"
  | "repair_query"
  | "verify_result"
  | "select_visualization"
  | "completed"
  | "clarification_required"
  | "out_of_scope"
  | "blocked"
  | "failed";

export interface GeneratedQuery {
  sql: string;
  expected_columns: string[];
}

export interface QueryResultColumn {
  name: string;
  type: "string" | "number" | "boolean" | "temporal" | "structured" | "unknown";
  semantic_type: "dimension" | "metric";
}

export interface QueryResult {
  columns: QueryResultColumn[];
  rows: unknown[][];
  row_count: number;
  truncated: boolean;
  duration_ms: number;
  warnings: string[];
}

export interface QueryRunError {
  code: string;
  message: string;
}

export interface QueryRun {
  run_id: string;
  datasource_id: string;
  question: string;
  status: QueryRunStatus;
  generated_query: GeneratedQuery | null;
  validation: Record<string, unknown>;
  result: QueryResult | null;
  repair_count: number;
  provider_call_count: number;
  visualization_type: string | null;
  clarification_suggestions: string[];
  warnings: string[];
  error: QueryRunError | null;
  created_at: string;
}

export interface TraceEvent {
  sequence: number;
  timestamp: string;
  state: QueryRunStatus;
  event: string;
  outcome: string;
  duration_ms?: number;
  error_code?: string;
  details?: Record<string, unknown>;
}

export interface QueryTrace {
  run_id: string;
  status: QueryRunStatus;
  trace: TraceEvent[];
}

export interface QueryRunSummary {
  run_id: string;
  datasource_id: string;
  datasource_name: string;
  question: string;
  status: QueryRunStatus;
  row_count: number | null;
  duration_ms: number | null;
  repair_count: number;
  provider_call_count: number;
  visualization_type: string | null;
  error_code: string | null;
  created_at: string;
}

export interface QueryRunPage {
  items: QueryRunSummary[];
  limit: number;
  total: number;
  next_cursor: string | null;
  has_more: boolean;
}

export interface QueryRunFilters {
  limit?: number;
  cursor?: string;
  status?: QueryRunStatus | "";
  datasourceId?: string;
  createdBefore?: string;
  search?: string;
}

export function createQueryRun(datasourceId: string, question: string) {
  return apiRequest<QueryRun>("/api/v1/query-runs", {
    method: "POST",
    body: JSON.stringify({ datasource_id: datasourceId, question }),
  });
}

export const getQueryRun = (runId: string) =>
  apiRequest<QueryRun>(`/api/v1/query-runs/${runId}`);

export const getQueryTrace = (runId: string) =>
  apiRequest<QueryTrace>(`/api/v1/query-runs/${runId}/trace`);

export function listQueryRuns(filters: QueryRunFilters = {}) {
  const params = new URLSearchParams();
  params.set("limit", String(filters.limit ?? 20));
  if (filters.cursor) params.set("cursor", filters.cursor);
  if (filters.status) params.set("status", filters.status);
  if (filters.datasourceId) params.set("datasource_id", filters.datasourceId);
  if (filters.createdBefore) params.set("created_before", filters.createdBefore);
  if (filters.search?.trim()) params.set("search", filters.search.trim());
  return apiRequest<QueryRunPage>(`/api/v1/query-runs?${params.toString()}`);
}
